specs/electra/beacon-chain.md

```diff
@@ -835,44 +873,116 @@ def process_registry_updates(state: BeaconState) -> None:
             # Deposited balance will never become active. Increase balance but do not consume churn
             else:
-                increase_balance(state, deposit.index, deposit.amount)
+                increase_balance(state, validator_index, deposit.amount)
         # Validator is not exiting, attempt to process deposit
         else:
-            # Deposit does not fit in the churn, no more deposit processing in this epoch.
-            if processed_amount + deposit.amount > available_for_processing:
-                break
-            # Deposit fits in the churn, process it. Increase balance and consume churn.
-            else: 
-                increase_balance(state, deposit.index, deposit.amount)
-                processed_amount += deposit.amount
+            # Increase balance
+            increase_balance(state, validator_index, deposit.amount)
+            # Check if valid deposit switch to compounding credentials.
+            if (
+                is_compounding_withdrawal_credential(deposit.withdrawal_credentials)
+                and has_eth1_withdrawal_credential(validator)
+                and is_valid_deposit_signature(
+                    deposit.pubkey,
+                    deposit.withdrawal_credentials,
+                    deposit.amount,
+                    deposit.signature
+                )
+            ):
+                switch_to_compounding_validator(state, validator_index)
+
+    return True
+```
+
+#### New `process_pending_deposits`
+
+Iterating over `pending_deposits` queue this function runs the following checks before applying pending deposit:
+1. All Eth1 bridge deposits are processed before the first deposit request gets processed.
+2. Deposit position in the queue is finalized.
+3. Deposit does not exceed the `MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING` limit.
+4. Deposit does not exceed the activation churn limit.
+
+```python
+def process_pending_deposits(state: BeaconState) -> None:
+    available_for_processing = state.deposit_balance_to_consume + get_activation_exit_churn_limit(state)
+    processed_amount = 0
+    next_deposit_index = 0
+    deposits_to_postpone = []
+    is_churn_limit_reached = False
+    finalized_slot = compute_start_slot_at_epoch(state.finalized_checkpoint.epoch)
+
+    for deposit in state.pending_deposits:
+        # Do not process deposit requests if Eth1 bridge deposits are not yet applied.
+        if (
+            # Is deposit request
+            deposit.slot > GENESIS_SLOT and
+            # There are pending Eth1 bridge deposits
+            state.eth1_deposit_index < state.deposit_requests_start_index
+        ):
+            break
+
+        # Check if deposit has been finalized, otherwise, stop processing.
+        if deposit.slot > finalized_slot:
+            break
+
+        # Check if number of processed deposits has not reached the limit, otherwise, stop processing.
+        if next_deposit_index > MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING:
```

Comment:
I think because we start at index 0 for next_deposit_index, `next_deposit_index > MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING` only triggers when  next_deposit_index has incremented MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING+2 times, which i'm not sure was the intended. maybe this needs to be
`next_deposit_index > MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING-1` to account for the 0 index?

Reply:
Good catch! It should rather be:
```suggestion
        if next_deposit_index >= MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING:
```